### DISCUSSION TITLE
Chaining async atomWithSuspenseQuery atoms from jotai-react-query

### DISCUSSION THREAD

**Discussion by sazzer** [pmndrs/jotai#2557] [Q&A]
Hi there,

I'm wondering if anyone can help? I'm using Jotai with React Query, and that's all working great. However, I'm trying to build a UI that works against a fully hypermedia API, so some of the calls I need to make can only be determined based on the result of another call. And, of course, all of the calls are asynchronous.

Writing an `atomWithSuspenseQuery` that makes a single API call is easy. However, what I'm needing to do is resolve a value from one `atomWithSuspenseQuery` for use in another one, and I can't work out how to correctly do this.

For example, in order to get the OpenID connect metadata, I need to:
* Call the API home document - This would be an `atomWithSuspenseQuery`.
* Extract a URL from this - which might or might not be present.
* Make another call to this URL - if present. This would be the `atomWithSuspenseQuery` that the UI component actually cares about.

Is there a good way to do this? The best I've been able to work out so far is to write an `atomFamily` that takes the URL as a parameter, and then call the two in succession from a custom hook. Which works, but feels a bit icky.

```
// home.ts
import { useAtom } from "jotai";
import { atomWithSuspenseQuery } from "jotai-tanstack-query";

import { doFetch } from "../api";

/** Atom for getting the API home document */
const apiHomeDocumentAtom = atomWithSuspenseQuery(() => ({
  queryKey: ["home"],
  queryFn: async () => {
    return await doFetch({ url: "/" });
  }
}));
apiHomeDocumentAtom.debugLabel = "app/home";

export function useApiHome() {
  const [home] = useAtom(apiHomeDocumentAtom);

  return home.data;
}
```

```
// oidc.ts

import { useAtom } from "jotai";
import { atomWithSuspenseQuery } from "jotai-tanstack-query";
import { atomFamily } from "jotai/vanilla/utils";

import { doFetch } from "../api";
import { useApiHome } from "../home";

type OidcMetadata = {
  authorization_endpoint: string;
};

const oidcMetadataAtomFamily = atomFamily((oidcMetadataUrl?: string) => {
  const oidcAtom = atomWithSuspenseQuery(() => ({
    queryKey: ["oidcMetadata", oidcMetadataUrl],
    queryFn: async () => {
      if (oidcMetadataUrl === undefined) {
        throw new Error("No URL to get OpenID Connect metadata");
      }

      return await doFetch({ url: oidcMetadataUrl });
    }
  }));
  oidcAtom.debugLabel = "app/oidc";

  return oidcAtom;
});

export function useOidcMetadata() {
  const homeDocument = useApiHome();

  const oidcMetadataUrl = homeDocument.getLink("oauth_server_metadata_uri")?.href;

  const [oidcMetadata] = useAtom(oidcMetadataAtomFamily(oidcMetadataUrl));

  return oidcMetadata.data.body;
}
```

Cheers

**Comment by dai-shi** [comment]
maybe @kalijonn ?

**Comment by kalijonn** [comment]
You can use dependent queries to conditionally query the api. However it doesn't work for atomWithSuspenseQuery yet. you can use atomWithQuery which has the enabled flag to conditionally query the api.

```ts
import { unwrap } from 'jotai/utils'

const oidcAtom = atomWithQuery((get) => {
  const homeResponse = get(unwrap(apiHomeDocumentAtom))

  const oidcMetadataUrl = homeResponse?.data.homeDocument.getLink("oauth_server_metadata_uri")?.href

  return {
    queryKey: ['oidcMetadata', oidcMetadataUrl],
    queryFn: async () => {
      if (oidcMetadataUrl === undefined) {
        throw new Error('No URL to get OpenID Connect metadata')
      }

      return await doFetch({ url: oidcMetadataUrl })
    },
    enabled: !! homeResponse,
  }
})
```
